daemon/cmd/ipcache.go

```diff
@@ -87,3 +104,126 @@ func containsSubnet(outer, inner net.IPNet) bool {
+		k := key.(*ipcachemap.Key)
+		v := value.(*ipcachemap.RemoteEndpointInfo)
+		nid := identity.NumericIdentity(v.SecurityIdentity)
+
+		if isLocalIdentity(nid) {
+			d.restoredCIDRs[k.Prefix()] = nid
+		} else if nid == identity.ReservedIdentityIngress && v.TunnelEndpoint.IsZero() {
+			ingressIPs = append(ingressIPs, k.Prefix())
+		}
+	})
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("error dumping ipcache: %w", err)
+	}
+	// DumpWithCallback() leaves the ipcache map open, must close before opened for
+	// parallel mode in Daemon.initMaps()
+	ipcachemap.IPCacheMap().Close()
+
+	// Now that the map is dumped,
+	// - upsert relevant metadata in to the ipcache
+	// - reserve all existing CIDR identities
+	// - add Ingress IPs to local Node.
+	metaUpdates := make([]ipcache.MU, 0, len(ingressIPs)+len(d.restoredCIDRs))
+	nidsToReserve := make([]identity.NumericIdentity, 0, len(d.restoredCIDRs))
+
+	for prefix, nid := range d.restoredCIDRs {
+		nidsToReserve = append(nidsToReserve, nid)
+		metaUpdates = append(metaUpdates, ipcache.MU{
+			Prefix:   prefix,
+			Source:   source.Restored,
+			Resource: restoredCIDRResource,
+			Metadata: []ipcache.IPMetadata{ipcachetypes.RequestedIdentity(nid)},
+		})
+	}
+	for _, prefix := range ingressIPs {
+		metaUpdates = append(metaUpdates, ipcache.MU{
+			Prefix:   prefix,
+			Source:   source.Restored,
+			Resource: ipcachetypes.NewResourceID(ipcachetypes.ResourceKindDaemon, "", "ingress"),
+			Metadata: []ipcache.IPMetadata{labels.LabelIngress},
+		})
+
+		// Set any restored ingress IPs back on the LocalNode object
+		d.nodeLocalStore.Update(func(n *node.LocalNode) {
+			addr := prefix.Addr()
+			if addr.Is4() {
+				n.IPv4IngressIP = addr.AsSlice()
+			} else {
+				n.IPv6IngressIP = addr.AsSlice()
+			}
+		})
+	}
+
+	// Insert the batched changes in to the ipcache.
+	// Even though the ipcache map hasn't been initialized yet, this is
+	// safe to do so, because the ipcache's apply controller is currently
+	// paused.
+	d.ipcache.IdentityAllocator.ReserveIdentities(nidsToReserve)
+	d.ipcache.UpsertMetadataBatch(metaUpdates...)
+
+	return nil
+}
+
+// releaseRestoredCIDRS removes the "stub" metadata that was inserted
+// in to the ipcache when local identities were restored.
+// Any identities actually in use will still exist after this.
+//
+// This first waits a grace period (default 10 minutes).
+// This grace period is needed when running on an external workload
+// where policy synchronization is not done via k8s. Also in k8s
+// case it is prudent to allow concurrent endpoint regenerations to
+// (re-)allocate the restored identities before we release them.
+func (d *Daemon) releaseRestoredCIDRs() {
+	defer func() {
+		// release the memory held by restored CIDRs
+		d.restoredCIDRs = nil
+	}()
+	if len(d.restoredCIDRs) == 0 {
+		return
+	}
+
+	time.Sleep(option.Config.IdentityRestoreGracePeriod)
+
+	log.Infof("Releasing metadata for %d restored CIDR identities", len(d.restoredCIDRs))
```

Comment:
If we're making this an Info level log, please use structured logging.

(Does this message make any sense to users, do we need to inform them? I guess this is the kind of thing that helps as a smoking gun if someone notices issues 10m after startup, but users may be confused what it means to release metadata... is that a good thing? bad thing? Why am I being told?)

Reply:
If we intend this to be more user-facing, we could rephrase it in terms of "releasing preallocated memory for identity restore" or something, at least then it's a bit less ambiguous.

Reply:
Right, we've had enough bugs over the years with this logic that I wanted to log *something* by default. I'll think of a better message. Something like "Removing identity reservations for restored CIDR identities".

Reply:
I think "Removing reservations" sounds a bit less alarming than "releasing metadata", so :+1: . I guess time will tell if it continues to be confusing.